### PR TITLE
perf: Ipc exec multiple paths

### PR DESCRIPTION
**Pull request by mickvangelderen**

### CODE REVIEW

**Comment by mickvangelderen** [comment]
This PR seems to pass now but won't after merging https://github.com/pola-rs/polars/pull/15065

File: crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs [hunk 1]
```diff
@@ -101,7 +227,7 @@ impl Executor for IpcExec {
         };
 
         let profile_name = if state.has_node_timer() {
-            let mut ids = vec![self.path.to_string_lossy().into()];
+            let mut ids = vec![self.paths[0].to_string_lossy().into()];
```
Comment (mickvangelderen) [inline crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs:230]: Taken from the parquet implementation. Should we just map each path instead?

File: crates/polars-lazy/src/scan/ipc.rs [hunk 2]
```diff
@@ -70,10 +86,10 @@ impl LazyFileListReader for LazyIpcReader {
...
         lf.opt_state.file_caching = true;
 
-        // it is a bit hacky, but this `with_row_index` function updates the schema
-        if let Some(row_index) = args.row_index {
-            lf = lf.with_row_index(&row_index.name, Some(row_index.offset))
-        }
+        // // it is a bit hacky, but this `with_row_index` function updates the schema
+        // if let Some(row_index) = args.row_index {
+        //     lf = lf.with_row_index(&row_index.name, Some(row_index.offset))
+        // }
```
Comment (mickvangelderen) [inline crates/polars-lazy/src/scan/ipc.rs:92]: Tests pass if this is removed, but can they be removed?

**Comment by ritchie46** [comment]
Hmm.. Something goes wrong with the windows tests.

**Review by ritchie46 — COMMENTED** [review]
Looks good overall. Left a few comments.

File: crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs [hunk 3]
```diff
@@ -49,49 +52,237 @@ impl IpcExec {
...
+        let row_limit = n_rows.unwrap_or(IdxSize::MAX);
+
+        // Used to determine the next file to open. This guarantees the order.
+        let path_index = AtomicUsize::new(0);
+        let row_counter = RwLock::new(ConsecutiveCountState::new(self.paths.len()));
+
+        let index_and_dfs = (0..self.paths.len())
+            .into_par_iter()
+            .map(|_| -> PolarsResult<(usize, DataFrame)> {
+                let index = path_index.fetch_add(1, Ordering::SeqCst);
```
[the remark this answers is not in the document]
Comment (mickvangelderen) [reply crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs:89]: All I know is that `Ordering::SeqCst` gives the strongest guarantees, so I'll trust you on this.

[the remark this answers is not in the document]
Comment (mickvangelderen) [reply crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs:89]: Reading https://doc.rust-lang.org/nomicon/atomics.html#relaxed helped.

File: crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs [hunk 4]
```diff
@@ -49,49 +52,237 @@ impl IpcExec {
...
+    row_counter: ConsecutiveCountState,
+    row_index: Option<&RowIndex>,
+    row_limit: IdxSize,
+    predicate: Option<&Arc<dyn PhysicalExpr>>,
+) -> PolarsResult<DataFrame> {
+    index_and_dfs.sort_unstable_by(|(a, _), (b, _)| a.cmp(b));
+
+    #[cfg(debug_assertions)]
+    {
+        assert!(
```
[the remark this answers is not in the document]
Comment (mickvangelderen) [reply crates/polars-lazy/src/physical_plan/executors/scan/ipc.rs:240]: Good point. Probably a left over from the computation not fitting in the `assert` expression initially.

**Comment by mickvangelderen** [comment]
> Hmm.. Something goes wrong with the windows tests.

Testing this idea to really fix the windows tests https://github.com/pola-rs/polars/pull/15191

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**